### ISSUE TITLE
[java] UncommentedEmptyConstructor consider annotations on Constructor

### ISSUE THREAD

**Issue by vanniktech** [pmd/pmd#357]
**Rule Set:**

```xml
<rule ref="rulesets/java/design.xml"/>
```

**Description:**

The code below will be flagged by PMD with:

```xml
<violation beginline="6" endline="8" begincolumn="11" endcolumn="3" rule="UncommentedEmptyConstructor" ruleset="Design" package="com.ovyapp.android.feature.calendar" class="MyClass" method="MyClass" externalInfoUrl="https://pmd.github.io/pmd-5.5.5/pmd-java/rules/java/design.html#UncommentedEmptyConstructor" priority="3">
Document empty constructor
</violation>
```

Maybe this rule could be relaxed when an annotation is applied on the constructor, especially if it's `javax.inject.Inject` since this annotation can be picked up by annotation processors such as Dagger 2 and let them know that they can build the `MyClass` class.

**Code Sample demonstrating the issue:**

```java
import javax.inject.Inject;

public class MyClass {
  @Inject MyClass() {

  }
}
```

**Running PMD 5.5.5 through:** *Gradle*

**Comment by jsotuyod** [comment]
@vanniktech it's been a long time since I last used Dagger 2, but if I recall correctly, annotating a constructor with `@Inject` is only necessary if a non-default constructor should be used.

On your example, provided that somewhere there is a requirement to inject a `MyClass` instance, it will manage to create one even without the annotation, or in this case, the whole constructor. Am I mistaken?

**Comment by vanniktech** [comment]
@jsotuyod It is necessary even when there's only the default constructor from Object.

Trying to inject MyClass like this won't work:

```java
public class MyClass {
}
```

**Comment by jsotuyod** [comment]
@vanniktech yeap, just figured it out myself.

> If your class has `@Inject`-annotated fields but no `@Inject`-annotated constructor, Dagger will inject those fields if requested, but will not create new instances. Add a no-argument constructor with the `@Inject` annotation to indicate that Dagger may create instances as well.

Your proposal sounds reasonable, given the annotation the reason of being for the empty constructor is clear and no comment should be necessary. I guess we could accept it as an enhancement.

**Comment by vanniktech** [comment]
@jsotuyod is someone working on this? Also do you guys accept PRs?

**Comment by ryan-gustafson** [comment]
PRs are accepted! You can find this via the main PMD site, but for your convenience:  https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md

**Comment by jsotuyod** [comment]
To be included in PMD 6.0.0
